lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -0,0 +1,160 @@
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newAllocOp;
+    if (hasParallelParent) {
+      newAllocOp = rewriter.create<cpuruntime::ThreadAllocOp>(
+          loc, type, dynamicSizes, symbolOperands);
+    } else {
+      newAllocOp = rewriter.create<cpuruntime::AllocOp>(loc, type, dynamicSizes,
+                                                        symbolOperands);
+    }
+    rewriter.replaceOp(op, newAllocOp->getResults());
+    return success();
+  }
+};
+
+struct AlignedDeallocLowering : public OpRewritePattern<memref::DeallocOp> {
+  using OpRewritePattern<memref::DeallocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::DeallocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    Value memref = op.getMemref();
+    // Check if the parent contains a forall / parallel loop
+    bool hasParallelParent = false;
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newDeallocOp;
+    if (hasParallelParent) {
+      newDeallocOp = rewriter.create<cpuruntime::ThreadDeallocOp>(loc, memref);
+    } else {
+      newDeallocOp = rewriter.create<cpuruntime::DeallocOp>(loc, memref);
+    }
+    rewriter.replaceOp(op, newDeallocOp->getResults());
+    return success();
+  }
+};
+
+struct ConvertMemRefToCPURuntime
+    : public impl::ConvertMemRefToCPURuntimeBase<ConvertMemRefToCPURuntime> {
+
+  void runOnOperation() final {
+    auto *ctx = &getContext();
+    // add lowering target
+    ConversionTarget target(getContext());
+    target.addLegalDialect<
+        // clang-format off
+        BuiltinDialect,
+        func::FuncDialect,
+        memref::MemRefDialect,
+        cpuruntime::CPURuntimeDialect,
+        math::MathDialect,
+        arith::ArithDialect,
+        scf::SCFDialect,
+        tensor::TensorDialect,
+        linalg::LinalgDialect,
+        affine::AffineDialect
+        // clang-format on
+        >();
+    target.addIllegalOp<memref::AllocOp, memref::DeallocOp>();
+    // set pattern
+    RewritePatternSet patterns(ctx);
+    patterns.add<AlignedAllocLowering>(ctx);
+    patterns.add<AlignedDeallocLowering>(ctx);
+
+    // Traverse the entire graph to find returnlikeOp operations
+    getOperation()->walk([this, ctx](Operation *op) {
+      if (op->hasTrait<OpTrait::ReturnLike>()) {
+        for (Value operand : op->getOperands()) {
+          if (operand.getType().isa<MemRefType>()) {
+            // Use buffer analysis to trace the source of the memref
+            Value memrefSrc = getViewBase(operand);
```

Comment:
Please use `BufferViewFlowAnalysis` instead of simply checking view.

Reply:
updated.